### PR TITLE
feat(wallet): add mojo method to query LiFi transaction status

### PR DESCRIPTION
**Pull request by onyb**
Resolves https://github.com/brave/brave-browser/issues/40100

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

N/A

### CODE REVIEW

File: components/brave_wallet/browser/swap_response_parser.cc [hunk 1]
```diff
@@ -379,6 +379,102 @@ std::optional<std::string> ChainIdToHex(const std::string& value) {
...
+    default:
+      return mojom::LiFiSubstatusCode::kUnknownError;
+  }
+}
+
+mojom::LiFiStepStatusPtr ParseStepStatus(
+    const swap_responses::LiFiStepStatus& value) {
+  auto result = mojom::LiFiStepStatus::New();
+
+  if (auto chain_id = ChainIdToHex(value.chain_id); chain_id) {
```
Comment (supermassive) [inline components/brave_wallet/browser/swap_response_parser.cc:461]: ~; chain_id~

File: components/brave_wallet/browser/swap_service.cc [hunk 2]
```diff
@@ -661,4 +666,40 @@ void SwapService::OnGetLiFiTransaction(GetTransactionCallback callback,
...
+  if (!api_request_result.Is2XXResponseCode()) {
+    if (auto error_response =
+            lifi::ParseErrorResponse(api_request_result.value_body())) {
+      std::move(callback).Run(nullptr, std::move(error_response), "");
+    } else {
+      std::move(callback).Run(
+          nullptr, nullptr, l10n_util::GetStringUTF8(IDS_WALLET_PARSING_ERROR));
+    }
+
+    return;
```
Comment (supermassive) [inline components/brave_wallet/browser/swap_service.cc:693]: nit: I'd suggest to pair `std::move(callback).Run(...)` with immediate `return`: it's a bit easier to review for missing returns and then possbile incorrect double runs of callbacks.

Though this file does not follow that.

File: components/brave_wallet/browser/swap_response_parser.cc [hunk 3]
```diff
@@ -751,6 +847,39 @@ mojom::LiFiErrorPtr ParseErrorResponse(const base::Value& json_value) {
...
+    result->sending = std::move(sending);
+  } else {
+    return nullptr;
+  }
+
+  if (auto receiving = ParseStepStatus(value->receiving); receiving) {
+    result->receiving = std::move(receiving);
+  } else {
+    return nullptr;
+  }
```
Comment (supermassive) [inline components/brave_wallet/browser/swap_response_parser.cc:869]: `if (auto x = something())`
having `; x` check is redundant in these ifs

**Review by supermassive — APPROVED** [review]
wallet core lgtm with some nits

**Review by nuo-xu — APPROVED** [review]
pushed some changes for iOS build. 
other lgtm

**Comment by goodov** [comment]
@onyb seeing this during the build, but the build seems like fine with that. Please fix to not trigger this message.

```
../../brave\components\brave_wallet\browser\swap_responses.idl(155) : Error : Trailing comma in block.
```

**Pull request merged**